File: hashes/3606-coupon-code-validator.py
import unittest
from typing import List, Tuple

# https://leetcode.com/problems/coupon-code-validator/

# python3 -m unittest hashes/3606-coupon-code-validator.py


class Solution(unittest.TestCase):
    # Approach: Filtering + Sorting with Custom Comparator
    # Time: O(n log n) - sorting valid coupons
    # Space: O(n) - storing valid coupons
    def validateCoupons(self, code: List[str], businessLine: List[str], isActive: List[bool]) -> List[str]:
        # Define business line priority order
        priority = {"electronics": 0, "grocery": 1, "pharmacy": 2, "restaurant": 3}
        valid_businesses = set(priority.keys())

        def is_valid_code(c: str) -> bool:
            """Check if code is non-empty and contains only alphanumeric chars and underscores"""
            if not c:
                return False
            return all(ch.isalnum() or ch == '_' for ch in c)

        # Collect valid coupons as (priority, code) tuples
        valid_coupons: List[Tuple[int, str]] = []
        for i in range(len(code)):
            if (isActive[i] and
                is_valid_code(code[i]) and
                businessLine[i] in valid_businesses):
                valid_coupons.append((priority[businessLine[i]], code[i]))

        # Sort by priority first, then by code lexicographically
        valid_coupons.sort()

        # Extract and return just the codes
        return [c for _, c in valid_coupons]

    def test(self):
        for code, businessLine, isActive, expected in [
            (
                ["SAVE20", "", "PHARMA5", "SAVE@20"],
                ["restaurant", "grocery", "pharmacy", "restaurant"],
                [True, True, True, True],
                ["PHARMA5", "SAVE20"],
            ),
            (
                ["GROCERY15", "ELECTRONICS_50", "DISCOUNT10"],
                ["grocery", "electronics", "invalid"],
                [False, True, True],
                ["ELECTRONICS_50"],
            ),
            (
                ["E1", "E2", "G1", "P1", "R1"],
                ["electronics", "electronics", "grocery", "pharmacy", "restaurant"],
                [True, True, True, True, True],
                ["E1", "E2", "G1", "P1", "R1"],
            ),
        ]:
            output = self.validateCoupons(code, businessLine, isActive)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")

File: hashes/3606-coupon-code-validator.go
package hashes

import (
	"sort"
	"unicode"
)

// https://leetcode.com/problems/coupon-code-validator/

// Approach: Filtering + Sorting with Custom Comparator
// Time: O(n log n) - sorting valid coupons
// Space: O(n) - storing valid coupons
func validateCoupons(code []string, businessLine []string, isActive []bool) []string {
	// Define business line priority order
	priority := map[string]int{
		"electronics": 0,
		"grocery":     1,
		"pharmacy":    2,
		"restaurant":  3,
	}

	isValidCode := func(c string) bool {
		// Check if code is non-empty and contains only alphanumeric chars and underscores
		if len(c) == 0 {
			return false
		}

		for _, ch := range c {
			if !unicode.IsLetter(ch) && !unicode.IsDigit(ch) && ch != '_' {
				return false
			}
		}

		return true
	}

	// Collect valid coupons as structs
	type Coupon struct {
		priority int
		code     string
	}

	var validCoupons []Coupon

	for i := range code {
		_, validBusiness := priority[businessLine[i]]
		if isActive[i] && isValidCode(code[i]) && validBusiness {
			validCoupons = append(validCoupons, Coupon{
				priority: priority[businessLine[i]],
				code:     code[i],
			})
		}
	}

	// Sort by priority first, then by code lexicographically
	sort.Slice(validCoupons, func(i, j int) bool {
		if validCoupons[i].priority != validCoupons[j].priority {
			return validCoupons[i].priority < validCoupons[j].priority
		}

		return validCoupons[i].code < validCoupons[j].code
	})

	// Extract and return just the codes
	result := make([]string, len(validCoupons))
	for i, coupon := range validCoupons {
		result[i] = coupon.code
	}

	return result
}
